### PR TITLE
AAP-48726: Fallback to controller auth

### PR DESCRIPTION
**Pull request by zkayyali812**
## Description
- What is being changed?
Updates the local authenticator to fallback to controller authentication when logging in via gateway local authenticator.
- Why is this change needed?
This change is needed to support users logging in in the following case -
1. A users AAP account was migrated over from 2.4->2.6 (this causes the account to exist in both AAP and gateway, but gateway is not aware of the user password, since we can't migrate over the password).
2. When a user attempts to login via gateway the initial auth fails, we check -
    a. If the user actually exists in gateway as a local user
    b. If the user is a valid user in controller, and we check if we can authenticate successfully via the `/api/controller/v2/me/` endpoint
3. If checks pass, we update the gateway user account with the updated password.

- How does this change address the issue?
This change addresses the issue by adding the necessary fallback logic.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist

- [ ] I have performed a self-review of my code
- [ ] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [ ] I have considered the security impact of these changes
- [ ] I have considered performance implications
- [ ] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

### Steps to Test
1. 
4. 
5. 

### Required Actions

- [ ] Requires documentation updates
  
- [ ] Requires downstream repository changes
  
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires coordination with other teams
  
- [ ] Blocked by PR/MR: #XXX

### CODE REVIEW

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 1]
```diff
@@ -69,5 +82,100 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        1. The user exists in the gateway database
+        2. The user is_partially_migrated=True
+        """
+        try:
+            user = UserModel._default_manager.get_by_natural_key(username)
+        except UserModel.DoesNotExist:
+            logger.warning(f"User '{username}' does not exist in the database.")
+            return False
+
+        if not user.resource.is_partially_migrated:
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]: I think `is_partially_migrated` is going away so we might not want to reference it here. If we remove this I'm not sure how I feel about this function. I kind of like that we are checking if they are in the DB but also, if they weren't we would just fail on the password update anyway.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 2]
```diff
@@ -69,5 +82,100 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        if controller_user := self._get_controller_user(username, password):
+            # Validate controller_user is a dictionary
+            if not isinstance(controller_user, dict):
+                logger.warning(f"Unable to authenticate user '{username}' with controller. Invalid result format.")
+                return False
+
+            # Validate controller_user has a ldap_dn field, if it is not None, then the user is a local user
+            ldap_dn = controller_user.get("ldap_dn")
+            if ldap_dn is None or ldap_dn != "":
+                logger.warning(f"User '{username}' is unable to be confirmed as a local user.")
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]: I think we also have to check for `password == "$encrypted$"` too to ensure they are local. The social auth provides will make the user serializer remove the password field.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 3]
```diff
@@ -54,6 +62,11 @@ def authenticate(self, request, username=None, password=None, **kwargs):
 
         user = super().authenticate(request, username, password, **kwargs)
 
+        if not user and not (request and request.path.startswith('/api/controller/')) and self.is_a_partially_migrated_controller_user(username, password):
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]: Not sure about this. I'm assuming you are doing this so you can make the request to `/api/controller/v2/me/`. But if you go to `https://<host>/api/controller/v2/instances/` and click login you will be taken to `https://34.207.163.197/api/controller/login` which actually looks like the gateway page and you can log in with gateway creds. So I think having this that open might prevent that from working? Its kind of a nice case that you are logging in for the first time to the API via a controller endpoint.

Comment (john-westcott-iv) [reply ansible_base/authentication/authenticator_plugins/local.py]: Also, this portion of DAB is made to be used generically and this would prevent that from happening (making it very AAP specific). It might be better to have some kind of variable that would kick this logic off.

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 1 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Updated to remove this check

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 2 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Updated to include this check

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 3 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: I updated to instead ensure the path starts with /api/controller/v2/me/ to consider this

Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: I updated further to only check on `/api/gateway/v1/login/`

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 4]
```diff
@@ -53,6 +61,10 @@ def authenticate(self, request, username=None, password=None, **kwargs):
             return None
 
         user = super().authenticate(request, username, password, **kwargs)
+        if not user and request and request.path.startswith('/api/gateway/v1/login/') and self._can_authenticate_from_controller(username, password):
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]: `request.path.startswith('/api/gateway/v1/login/')` feels wrong, I wonder if we can make it less so. Could you try the below?

```
from django.urls import reverse
login_url = reverse('login')
```

Ideally we would somehow scope it to the Gateway app.

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 3 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Updated even further to be - 
```
and request.path.startswith(reverse('login', current_app='aap_gateway_api'))
```
Based on @chrismeyersfsu suggestion

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 4 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Hey Chris,
Great note. I updated this logic to be - 
```
and request.path.startswith(reverse('login', current_app='aap_gateway_api'))
```
to encapsulate that

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 5]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        """
+        Check if a user exists in the AuthenticatorUser table with the local authenticator provider.
+        If the user is valid, update the gateway users credentials with the controller credentials.
+        """
+        try:
+            user = UserModel._default_manager.get_by_natural_key(username)
+        except UserModel.DoesNotExist:
+            logger.warning(f"User '{username}' does not exist in the database.")
+            return False
+
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py:110]:
```suggestion
if user.role.is_partially_migrated is False:
  return False
```

I expected to find a short-circuit before calling to the Controller API.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 6]
```diff
@@ -53,6 +62,15 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
         user = super().authenticate(request, username, password, **kwargs)
+        if (
+            not user
+            and request
+            and request.path.startswith(reverse('login', current_app='aap_gateway_api'))
+            and self._can_authenticate_from_controller(username, password)
+        ):
+            logger.warning("User has been validated by controller, updating gateway user.")
+            self.update_gateway_user(username, password)
+            user = super().authenticate(request, username, password, **kwargs)
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py:73]:
```suggestion
            user = super().authenticate(request, username, password, **kwargs)
        else:
            logger.info(
                "Fallback authentication condition not met: "
                f"user={user}, "
                f"request={'set' if request else 'None'}, "
                f"login_path={'True' if request and request.path.startswith(reverse('login', current_app='aap_gateway_api')) else 'False'}, "
                f"can_auth_controller={'True' if self._can_authenticate_from_controller(username, password) else 'False'}"
            )
```

* So we, and an aap admin, can answer the quesiton: "why did my login fail?"
* info level (vs. debug level) so that, as an aap admin logging data to splunk, when there is an incident I can provide the security team with data about failed login attempts, by default.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 7]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        """
+
+        controller_base_domain = get_setting('gateway_proxy_url')
+        if not controller_base_domain:
+            logger.warning("Controller authentication failed, unable to get controller base domain")
+            return None
+        controller_url = urljoin(controller_base_domain, "/api/controller/v2/me/")
+
+        try:
+            response = requests.get(controller_url, auth=HTTPBasicAuth(username, password), timeout=10)
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]: TL;DR set the timeout to the underlying component request/response cycle timeout.

What settings are available to us related to request timeouts?
@kdelee 

You had to pick something, but I'd rather the timeout be calculated based on the remaining time left in the request/response cycle.

```
envoy --> grpc --> gateway --> controller
```
@kdelee has added a parameter to the installer, I can't remember the name of it, but it's the total request response time. That is the time envoy gets set to. Every other service to the right of envoy gets a timeout set that is < envoy, and that number gets subtracted from the request response time total. So you get `total request response time (envoy) = grpc + gateway + component`.

I'm not sure how we get a handle on that installer setting.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 8]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            # Check if count exists and equals 1
+            count = user_data.get("count")
+            if count != 1:
+                logger.warning(f"Unable to authenticate user '{username}' with controller.")
+                return None
+
+            # Check if results exists and is a non-empty list
+            results = user_data.get("results")
+            if not results or not isinstance(results, list) or len(results) == 0:
+                logger.warning(f"Unable to authenticate user '{username}' with controller. Invalid or empty results.")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
                logger.info(f"Unable to authenticate user '{username}' with controller. Invalid or empty results.")
```

I'm thinking info log for this one. This one can happen if someone types the wrong password. It's not necessarily a system issue

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 9]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+                return None
+            return results[0]
+        except requests.exceptions.HTTPError as http_err:
+            logger.info(f"HTTP error occurred: {http_err}")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            logger.warning(f"HTTP error occurred: {http_err}")
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 10]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+        except requests.exceptions.ConnectionError as conn_err:
+            logger.info(f"Connection error occurred: {conn_err}")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            logger.warning(f"Connection error occurred: {conn_err}")
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 11]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+        except requests.exceptions.Timeout as timeout_err:
+            logger.info(f"Timeout error occurred: {timeout_err}")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            logger.warning(f"Timeout error occurred: {timeout_err}")
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 12]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+        except requests.exceptions.RequestException as err:
+            logger.info(f"An unexpected error occurred: {err}")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            logger.warning(f"An unexpected error occurred: {err}")
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 13]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+        except ValueError as json_err:
+            logger.info(f"JSON decode error occurred: {json_err}")
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            logger.warning(f"JSON decode error occurred: {json_err}")
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 14]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+
+    def update_gateway_user(self, username, password):
+        """
+        Update the gateway user with the controller credentials and set is_partially_migrated to False.
+        """
+        user = UserModel._default_manager.get_by_natural_key(username)
+        user.set_password(password)
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py:185]:
```suggestion
        user.set_password(password)
        user.is_partially_migrated = False
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 15]
```diff
@@ -69,5 +87,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        user.save()
```
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
        user.save(update_fields=['password', 'is_partially_migrated'])
```

What I imagine the sql of the two looks like:

save sql: `UPDATE "auth_user" SET "username" = 'newusername', "email" = 'existing_email', "password" = 'existing_password' WHERE "auth_user"."id" = 1;`

`update_fields` sql: `UPDATE "auth_user" SET  "password" = 'password-hash-here', "is_partially_migrated" = False WHERE "auth_user"."id" = 1;`

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 13 above)
Comment (chrismeyersfsu) [inline ansible_base/authentication/authenticator_plugins/local.py]: Should we "blindly" include error messages from another service in our logs, especially when it's the authentication path?

Part of me likes it. The downstream service is ours, it shouldn't be outputting sensitive data in the response.

But then part of me thinks about layers of security/checks.

I vote for debuggability and to continue forwarding the message directly to the logs. No change.

**Story time:**
Controller recently had a security issue for doing this sort of thing with Cyberark responses or it was the GET url issued to Cyberark. Because Cyberark did a bad best practice (putting sensitive data in GET params) we had to account for it.

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 5 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:110]: This was removed due to @john-westcott-iv's comment here - https://github.com/ansible/django-ansible-base/pull/758#discussion_r2205103814

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 14 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:185]: is_partially_migrated was removed from this PR as this is likely going away

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 15 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Nice!! Did not know you could do this

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 13 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: I think this should be acceptable. I don't think any sensitive data is returned from this API directly. Even if a password is returned, it would be its encrypted value.

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 6 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:73]: I think Ill change the case here to `elif not user` so it does not trigger on every login

Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:73]: Also Ill change `user` to `username`

Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:73]: I also removed `_can_authenticate_from_controller` from this info log. If this is included, it adds risk to somehow lead to infinite loops if all the potential authentication paths are not excluded.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 16]
```diff
@@ -53,6 +62,22 @@ def authenticate(self, request, username=None, password=None, **kwargs):
             return None
 
         user = super().authenticate(request, username, password, **kwargs)
+        if (
+            not user
+            and request
+            and request.path.startswith(reverse('login', current_app='aap_gateway_api'))
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]: If this was ever run under a service w/o this endpoint this would raise an exception:
```
In [3]: reverse('login2', current_app='aap_gateway_api')
---------------------------------------------------------------------------
NoReverseMatch                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 reverse('login2', current_app='aap_gateway_api')

File ~opt/aap_gateway/venv/lib64/python3.11/site-packages/django/urls/base.py:88, in reverse(viewname, urlconf, args, kwargs, current_app)
     83     if ns_pattern:
     84         resolver = get_ns_resolver(
     85             ns_pattern, resolver, tuple(ns_converters.items())
     86         )
---> 88 return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)

File ~opt/aap_gateway/venv/lib64/python3.11/site-packages/django/urls/resolvers.py:828, in URLResolver._reverse_with_prefix(self, lookup_view, _prefix, *args, **kwargs)
    823 else:
    824     msg = (
    825         "Reverse for '%(view)s' not found. '%(view)s' is not "
    826         "a valid view function or pattern name." % {"view": lookup_view_s}
    827     )
--> 828 raise NoReverseMatch(msg)

NoReverseMatch: Reverse for 'login2' not found. 'login2' is not a valid view function or pattern name.

```

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 6 above)
Comment (john-westcott-iv) [reply ansible_base/authentication/authenticator_plugins/local.py:73]: I've never tried it in an and condition but could you do:
```
and controller_login_results := self._can_authenticate_from_controller(username, password)
```

And then in the print statement just dump the variable?

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 5 above)
Comment (john-westcott-iv) [reply ansible_base/authentication/authenticator_plugins/local.py:110]: Yea, we can double check with the upgrade patch but I'm pretty sure removing this flag was part of that. @AlanCoding might be able to confirm off the top of his head.

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 17]
```diff
@@ -69,5 +94,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+                logger.warning(f"Unable to authenticate user '{username}' with controller. Invalid result format.")
+                return False
+            # Validate controller_user has a ldap_dn field, if it is not None, then the user is a local user
+            ldap_dn = controller_user.get("ldap_dn")
+            if (ldap_dn is None or ldap_dn != "") and user.password == "$encrypted$":
+                logger.warning(f"User '{username}' is unable to be confirmed as a local user.")
+                return False
+            return True
+        else:
+            return False
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]:
```suggestion
            # We can only validate local controller users so we need to check two things
            # First lets see if they were an ldap user
            ldap_dn = controller_user.get("ldap_dn", None)
            if ldap_dn is None or ldap_dn != "":
                logger.warning(f"User '{username}' is an ldap user and can not be authenticated.")
                return False
             # Second, lets see if the were an enterprise user
             if controller_user.get('password, None) != "$encrypted$":
                logger.warning(f"User '{username}' is an enterprise user and can not be authenticated.")
                return False
            return True
        else:
            return False
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 18]
```diff
@@ -69,5 +94,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            return None
+        except requests.exceptions.Timeout as timeout_err:
+            logger.warning(f"Timeout error occurred: {timeout_err}")
+            return None
+        except requests.exceptions.RequestException as err:
+            logger.warning(f"An unexpected error occurred: {err}")
+            return None
+        except ValueError as json_err:
+            logger.warning(f"JSON decode error occurred: {json_err}")
+            return None
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py:175]: Do we also want to catch a generic Exception here incase something else is raised up?

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 19]
```diff
@@ -69,5 +94,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            if count != 1:
+                logger.warning(f"Unable to authenticate user '{username}' with controller.")
+                return None
+
+            # Check if results exists and is a non-empty list
+            results = user_data.get("results")
+            if not results or not isinstance(results, list) or len(results) == 0:
+                logger.info(f"Unable to authenticate user '{username}' with controller. Invalid or empty results.")
+                return None
+            return results[0]
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py:160]:
```suggestion
            if not isinstance(results[0], dict):
                logger.warning(f"Unable to authenticate user '{username}' with controller. user was not a dictionary.")
                return False
                
            return results[0]
```

File: ansible_base/authentication/authenticator_plugins/local.py [hunk 20]
```diff
@@ -69,5 +94,84 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+            user = UserModel._default_manager.get_by_natural_key(username)
+        except UserModel.DoesNotExist:
+            logger.warning(f"User '{username}' does not exist in the database.")
+            return False
+
+        if controller_user := self._get_controller_user(username, password):
+            # Validate controller_user is a dictionary
+            if not isinstance(controller_user, dict):
+                logger.warning(f"Unable to authenticate user '{username}' with controller. Invalid result format.")
+                return False
```
Comment (john-westcott-iv) [inline ansible_base/authentication/authenticator_plugins/local.py]: I almost want to move this into `_get_controller_user` so that you are dealing with entirety of the response in that function. (i.e. because you already check the results count and what not in there). But this is a nit. 

```suggestion
```

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 5 above)
Comment (chrismeyersfsu) [reply ansible_base/authentication/authenticator_plugins/local.py:110]: Asked if `is_partially_migrated` is going away slack https://redhat-internal.slack.com/archives/C08NZHY16Q1/p1752586713478919

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 7 above)
Comment (kdelee) [reply ansible_base/authentication/authenticator_plugins/local.py]: Do you have access to the gateway Django settings? 

```
{
    "gateway_token_name": "X-DAB-JW-TOKEN",
    "gateway_access_token_expiration": 600,
    "jwt_expiration_buffer_in_seconds": 2,
    "gateway_basic_auth_enabled": true,
    "gateway_proxy_url": "http://elijah-minimal-25.aap-25.svc.cluster.local/",
    "gateway_proxy_url_ignore_cert": false,
    "jwt_private_key": "$encrypted$",
    "jwt_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4AP9e+qu6JRGJYRm2YxN\nXqJmzpCxrffiInidINfcLwbQ787y4S+xI7IH5iVsjfiGSS6Q8weY8WIIh/vvJ8fz\n2mH+9mfFBpmR8/CYaFtzQvaLnYSGt8EmlDzDLnJ0rQTugVi6V7gt0OzU+6cAv1ej\nwbBfYM/tSBp39i5O3JV4aFkIz4G0HPM9ghLwuXM4I66+uBs131iTCKuZtIlUK5Xl\n54K4j7ubrW5b47hGCsvp/HO4mjS6aYH04gqvPdsfiEjdtmAAVC2yNR3dt4DuKBDi\n+vjZSPXXSpyhndTrIoWhegS9lP9AU4WT6NBEFvqz63nb6Ge6Pic+tB5FRs+1atnD\nZwIDAQAB\n-----END PUBLIC KEY-----\n",
    "status_endpoint_backend_timeout_seconds": 5,
    "status_endpoint_backend_verify": true,
    "resource_client_request_timeout": 2.0,
    "request_timeout": 30,
    "password_min_length": 0,
    "password_min_digits": 0,
    "password_min_upper": 0,
    "password_min_special": 0,
    "allow_admins_to_set_insecure": false,
    "LOGIN_REDIRECT_OVERRIDE": "",
    "custom_login_info": "",
    "custom_logo": "",
    "SOCIAL_AUTH_USERNAME_IS_FULL_EMAIL": false,
    "SESSION_COOKIE_AGE": 900,
    "DEFAULT_PAGE_SIZE": 50,
    "MAX_PAGE_SIZE": 200,
    "ALLOW_OAUTH2_FOR_EXTERNAL_USERS": false,
    "CONTROLLER_SSO_URL": "",
    "AUTOMATION_HUB_SSO_URL": "",
    "INSIGHTS_TRACKING_STATE": true,
    "RED_HAT_CONSOLE_URL": "https://console.redhat.com/",
    "REDHAT_USERNAME": "",
    "REDHAT_PASSWORD": "$encrypted$",
    "SUBSCRIPTIONS_USERNAME": "",
    "SUBSCRIPTIONS_PASSWORD": "$encrypted$",
    "AUTOMATION_ANALYTICS_GATHER_INTERVAL": 14400
}
```
Gateway's  setting `request_timeout` is the timeout for the components (their nginx + wsgi) because grpc timeout is counted separately 
```
settings.GRPC_SERVER_AUTH_SERVICE_TIMEOUT
'30s'
```
This is the grpc service timeout.

The top level `client_request_timeout` is `request_timeout + settings.GRPC_SERVER_AUTH_SERVICE_TIMEOUT` but is not set anywhere in gateway settings

Controller/EDA/Hub don't seem to have any setting INSIDE the application that makes it aware of the nginx/uwsgi/gunicorn timeouts

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 16 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: In this case I am going to remove the `reverse` and go back to a hardcoded url. 
CC: @chrismeyersfsu . The other alternative is to do a try/catch to try to get this value, but for simplicity, probably easier to just hardcode the expected path.

Open to other options/opinions here though

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 5 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:110]: Adding a note here for posterity - it was confirmed by @AlanCoding that `is_partially_migrated` is going away

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 14 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:185]: Noting for posterity, that it was confirmed in slack that this is going away

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 5 above)
Comment (AlanCoding) [reply ansible_base/authentication/authenticator_plugins/local.py:110]: yep ping @TheRealHaoLiu

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 6 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:73]: Done

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 7 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Added timeout based on - `GRPC_SERVER_AUTH_SERVICE_TIMEOUT`. Will fallback to `30` seconds if not provided

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 18 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:175]: Added

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 19 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py:160]: Done

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 17 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Done

File: ansible_base/authentication/authenticator_plugins/local.py (hunk 20 above)
Comment (zkayyali812) [reply ansible_base/authentication/authenticator_plugins/local.py]: Done
